ql/src/java/org/apache/hadoop/hive/ql/optimizer/UnionDistinctMerger.java

```diff
@@ -0,0 +1,181 @@
+import java.util.HashSet;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Stack;
+
+public class UnionDistinctMerger extends Transform {
+  private static final Logger LOG = LoggerFactory.getLogger(UnionDistinctMerger.class);
+
+  private static final String patternString = new StringBuilder()
+      .append(UnionOperator.getOperatorName()).append("%")
+      .append(GroupByOperator.getOperatorName()).append("%")
+      .append(ReduceSinkOperator.getOperatorName()).append("%")
+      .append(GroupByOperator.getOperatorName()).append("%")
+      .append(UnionOperator.getOperatorName()).append("%")
+      .append(GroupByOperator.getOperatorName()).append("%")
+      .append(ReduceSinkOperator.getOperatorName()).append("%")
+      .append(GroupByOperator.getOperatorName()).append("%")
+      .toString();
+
+  private static class UnionMergeContext implements NodeProcessorCtx {
+    public final ParseContext pCtx;
+
+    public UnionMergeContext(ParseContext pCtx) {
+      this.pCtx = pCtx;
+    }
+  }
+
+  private class UnionMergeProcessor implements SemanticNodeProcessor {
+    @Override
+    public Object process(Node nd, Stack<Node> stack, NodeProcessorCtx procCtx,
+        Object... nodeOutputs) throws SemanticException {
+      UnionMergeContext context = (UnionMergeContext) procCtx;
+
+      // The stack contains at least 8 operators, UNION-GBY-RS-GBY-UNION-GBY-RS-GBY.
+      // The leftmost UNION is on stack.size() - 8 and the rightmost GBY is on stack.size() - 1.
+      HashSet<Operator> allOps = new HashSet<>(context.pCtx.getAllOps());
+      for (int i = 1; i <= 8; i ++) {
+        Operator<?> op = (Operator<?>) stack.get(stack.size() - i);
+
+        // We do not apply the optimization if some operators do not belong to query plan.
+        // This can be happened when we already merged some UNIONs before.
+        // For example, Suppose that a query plan looks like the below graph:
+        //   (1)UNION-GBY-RS-GBY-(3)UNION-GBY-RS-GBY
+        //   (2)UNION-GBY-RS-GBY-
+        // Then we merge (1)-(3) to (1) and them move on to merging (2)-(3). Without checking the presence of
+        // operators in (2) and (3), the merge process will fail as (3) is already removed.
+        if (!allOps.contains(op)) {
+          return null;
+        }
+
+        // We do not apply the optimization if intermediate outputs are used by other operators.
+        if (i != 1 && op.getChildOperators().size() > 1) {
+          return null;
+        }
+      }
+
+      UnionOperator upperUnionOperator = (UnionOperator) stack.get(stack.size() - 8);
+      GroupByOperator upperFinalGroupByOperator = (GroupByOperator) stack.get(stack.size() - 5);
+
+      UnionOperator lowerUnionOperator = (UnionOperator) stack.get(stack.size() - 4);
+      GroupByOperator lowerFinalGroupByOperator = (GroupByOperator) stack.get(stack.size() - 1);
+
+      if (upperFinalGroupByOperator.getConf().getAggregators().isEmpty() &&
```

Comment:
could you please add a comment on what it means if `aggregators().isEmpty()`

Reply:
added a comment about `aggregators().isEmpty()`